File: python/54_SpiralMatrix.py
from unittest import TestCase
# https://leetcode.com/problems/spiral-matrix/


class SpiralMatrix(TestCase):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        result = []
        if not matrix: return result

        steps = [len(matrix[0]), len(matrix)]

        i, j = 0, -1
        dir = [[0,1], [1,0], [0,-1], [-1,0]]
        d = 0

        while steps[d & 1] > 0:
            di, dj = dir[d & 3]
            for _ in range(steps[d & 1]):
                i += di
                j += dj
                result.append(matrix[i][j])
            d += 1
            steps[d & 1] -= 1
        return result

    def spiralOrder1(self, matrix):         # ...haha
        result = []
        while matrix:
            result += matrix.pop(0)
            matrix = list(zip(*matrix))[::-1]
        return result

    def test1(self):
        self.assertEqual([1,2,3,6,9,8,7,4,5], self.spiralOrder([
            [ 1, 2, 3 ],
            [ 4, 5, 6 ],
            [ 7, 8, 9 ]
        ]))

    def test2(self):
        self.assertEqual([1,2,3,4,8,12,11,10,9,5,6,7], self.spiralOrder([
            [1, 2, 3, 4],
            [5, 6, 7, 8],
            [9,10,11,12]
        ]))

File: kotlin/src/main/kotlin/54.SpiralMatrix.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import kotlin.arrayOf as arr
import kotlin.intArrayOf as iarr
//https://leetcode.com/problems/spiral-matrix/

class SpiralMatrix{
  fun spiralOrder(matrix: Array<IntArray>): List<Int> {
    val result = mutableListOf<Int>()
    if(matrix.isEmpty()) return result

    var turn=0
    val steps = mutableListOf(matrix[0].size, matrix.size)

    var i=0
    var j=-1
    var di=0
    var dj=1

    while(steps[turn] > 0){
      (1..steps[turn]).forEach{
        i += di
        j += dj
        result.add(matrix[i][j])
      }
      turn = turn xor 1
      steps[turn] --
      val tmp=di
      di = dj
      dj = -tmp
    }

    return result
  }

  @Test
  fun test1(){
    assertEquals(listOf(1,2,3,6,9,8,7,4,5), spiralOrder(arr(
        iarr(1, 2, 3),
        iarr(4, 5, 6),
        iarr(7, 8, 9)
    )))
  }

  @Test
  fun test2(){
    assertEquals(listOf(1,2,3,4,8,12,11,10,9,5,6,7), spiralOrder(arr(
        iarr(1, 2, 3, 4),
        iarr(5, 6, 7, 8),
        iarr(9,10,11,12)
    )))
  }
}
